src/Nethermind/Nethermind.TxPool/TxPool.cs

```diff
@@ -338,6 +312,185 @@ private AddTxResult AddCore(Transaction tx, bool isPersistentBroadcast, bool isR
+            {
+                _hashCache.Set(tx.Hash);
+            }
+
+            /*
+             * we need to make sure that the sender is resolved before adding to the distinct tx pool
+             * as the address is used in the distinct value calculation
+             */
+            if (!isKnown)
+            {
+                lock (_locker)
+                {
+                    tx.GasBottleneck = tx.CalculateEffectiveGasPrice(_specProvider.GetSpec().IsEip1559Enabled, CurrentBaseFee);
+                    bool inserted = _transactions.TryInsert(tx.Hash, tx, out Transaction? removed);
+                    if (inserted)
+                    {
+                        UpdateBucket(tx.SenderAddress!);
+                        if (removed?.Hash is not null)
+                        {
+                            // transaction which was on last position in sorted TxPool and was deleted to give
+                            // a place for a newly added tx (with higher priority) is now removed from hashCache
+                            // to give it opportunity to come back to TxPool in the future, when fees drops
+                            _hashCache.Delete(removed.Hash);
+                        }
+                    }
+                    isKnown |= !inserted;
+                }
+            }
+
+            if (!isKnown)
+            {
+                isKnown |= !isReorg && (_txStorage.Get(tx.Hash) is not null);
+            }
+
+            if (isKnown)
+            {
+                // If transaction is a bit older and already known then it may be stored in the persistent storage.
+                Metrics.PendingTransactionsKnown++;
+                if (_logger.IsTrace) _logger.Trace($"Skipped adding transaction {tx.ToString("  ")}, already known.");
+                return AddTxResult.AlreadyKnown;
+            }
+            
+            HandleOwnTransaction(tx, isPersistentBroadcast);
+            NotifySelectedPeers(tx);
+            StoreTx(tx);
+            NewPending?.Invoke(this, new TxEventArgs(tx));
+            return AddTxResult.Added;
+        }
+
+        private void UpdateBucket(Address senderAddress)
+        {
+            var bucketSnapshot = _transactions.GetBucketSnapshot(senderAddress);
+            
+            if (bucketSnapshot.Length == 0)
+            {
+                return;
+            }
+            
+            Account? account = _stateProvider.GetAccount(senderAddress);
+            UInt256 balance = account?.Balance ?? UInt256.Zero;
+            long currentNonce = (long)(account?.Nonce ?? UInt256.Zero);
+
+            UpdateGasBottleneck(bucketSnapshot, currentNonce, balance);
+        }
+
+        private void UpdateGasBottleneck(IReadOnlyList<Transaction> bucketSnapshot, long currentNonce, UInt256 balance)
+        {
+            Transaction tx = bucketSnapshot[0];
+            UInt256 previousTxBottleneck = tx.IsEip1559 ? CalculatePayableGasPrice(tx, balance) : tx.GasPrice;
```

Comment:
Minor: move this condition statement to CalculatePayableGasPrice. Move CalculatePayableGasPrice to TransactionExtensions

Reply:
And then you can add unit tests for CalculatePayableGasPrice

Reply:
Good idea